executor/src/witgen/data_structures/identity.rs

```diff
@@ -0,0 +1,244 @@
+                    // Multiple matches
+                    Some(_) => {
+                        return None;
+                    }
+                }
+            }
+        }
+
+        matching_receive
+    }
+
+    fn interaction_id(&self) -> Option<T> {
+        match &self.selected_tuple.expressions[0] {
+            AlgebraicExpression::Number(id) => Some(*id),
+            _ => None,
+        }
+    }
+
+    pub fn is_unconstrained(&self) -> bool {
+        // TODO: This is a hack (but should work if it was originally a lookup / permutation)
+        self.multiplicity.as_ref() == Some(&self.selected_tuple.selector)
+    }
+}
+
+impl<T> Children<AlgebraicExpression<T>> for BusInteractionIdentity<T> {
+    fn children_mut(&mut self) -> Box<dyn Iterator<Item = &mut AlgebraicExpression<T>> + '_> {
+        Box::new(
+            self.selected_tuple
+                .children_mut()
+                .chain(self.multiplicity.iter_mut()),
+        )
+    }
+    fn children(&self) -> Box<dyn Iterator<Item = &AlgebraicExpression<T>> + '_> {
+        Box::new(
+            self.selected_tuple
+                .children()
+                .chain(self.multiplicity.iter()),
+        )
+    }
+}
+
+impl<T: fmt::Display + fmt::Debug> fmt::Display for BusInteractionIdentity<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "BusInteractionIdentity(id={}, multiplicity={:?}, tuple={})",
+            self.id, self.multiplicity, self.selected_tuple
+        )
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, derive_more::Display)]
+pub enum Identity<T> {
+    Polynomial(PolynomialIdentity<T>),
+    Connect(ConnectIdentity<T>),
+    BusSend(BusInteractionIdentity<T>),
+    BusReceive(BusInteractionIdentity<T>),
+}
+
+impl<T> Children<AlgebraicExpression<T>> for Identity<T> {
+    fn children_mut(&mut self) -> Box<dyn Iterator<Item = &mut AlgebraicExpression<T>> + '_> {
+        match self {
+            Identity::Polynomial(i) => i.children_mut(),
+            Identity::Connect(i) => i.children_mut(),
+            Identity::BusSend(i) | Identity::BusReceive(i) => i.children_mut(),
+        }
+    }
+
+    fn children(&self) -> Box<dyn Iterator<Item = &AlgebraicExpression<T>> + '_> {
+        match self {
+            Identity::Polynomial(i) => i.children(),
+            Identity::Connect(i) => i.children(),
+            Identity::BusSend(i) | Identity::BusReceive(i) => i.children(),
+        }
+    }
+}
+
+impl<T> Identity<T> {
+    pub fn contains_next_ref(&self) -> bool {
+        self.children().any(|e| e.contains_next_ref())
+    }
+
+    pub fn id(&self) -> u64 {
+        match self {
+            Identity::Polynomial(i) => i.id,
+            Identity::Connect(i) => i.id,
+            Identity::BusSend(i) | Identity::BusReceive(i) => i.id,
+        }
+    }
+}
+
+pub fn convert<T: Clone>(identities: &[AnalyzedIdentity<T>]) -> Vec<Identity<T>> {
+    let max_poly_id = identities
```

Comment:
Instead of doing these id gymnastics, can we reassign all ids at the end of this function based on the indices in the full vector?

Reply:
Yeah, that's probably better!